docs/spec/spec.md

```diff
@@ -0,0 +1,345 @@
+
+```ballerina
+service /ws on new websocket:Listener(21003) {
+    resource function get .(http:Request req) returns websocket:Service|websocket:UpgradeError {
+        return new WsService();
+}
+        
+service class WsService {
+  *websocket:Service;
+  remote isolated function onTextMessage(websocket:Caller caller, string data) returns websocket:Error? {
+      check caller->writeTextMessage(data);
+  }
+}              
+```
+
+#### Remote methods associated with websocket:Service
+
+onOpen: As soon as the WebSocket handshake is completed and the connection is established, the `onOpen` remote method is dispatched.
+```ballerina
+remote function onOpen(websocket:Caller caller) returns error? {
+   io:println("Opened a WebSocket connection"`);
+}
+```
+
+onTextMessage: The received text messages are dispatched to this remote method.
+```ballerina
+remote isolated function onTextMessage(websocket:Caller caller, string text) returns websocket:Error? {
+    io:println("Text message: " + text);
+}
+```
+
+onBinaryMessage: The received binary messages are dispatched to this remote method.
+```ballerina
+remote isolated function onBinaryMessage(websocket:Caller caller, byte[] data) returns websocket:Error? {
+    io:println("Binary message: " + data);
+}
+```
+
+onPing and onPong: The received ping and pong messages are dispatched to these remote methods respectively. You do not need to explicitly control these messages as they are handled automatically by the services and clients.
+```ballerina
+remote function onPing(websocket:Caller caller, byte[] data) returns error? {
+    io:println(string `Ping received with data: ${data.toBase64()}`);
+    check caller->pong(data);
+}
+ 
+remote function onPong(websocket:Caller caller, byte[] data) {
+    io:println(string `Pong received with data: ${data.toBase64()}`);
+}
+```
+
+onIdleTimeout: This remote method is dispatched when the idle timeout is reached. The idleTimeout has to be configured either in the WebSocket service or the client configuration.
+```ballerina
+remote function onIdleTimeout(websocket:Client caller) {
+    io:println("Connection timed out");
+}
+```
+
+onClose: This remote method is dispatched when a close frame with a statusCode and a reason is received.
+```ballerina
+remote function onClose(websocket:Caller caller, int statusCode, string reason) {
+    io:println(string `Client closed connection with ${statusCode} because of ${reason}`);
+}
+```
+
+onError: This remote method is dispatched when an error occurs in the WebSocket connection. This will always be preceded by a connection closure with an appropriate close frame.
+```ballerina
+remote function onError(websocket:Caller caller, error err) {
+    io:println(err.message());
+}
+```
+
+## 4. [Client](#4-client)
+
+`websocket:Client` can be used to send and receive data synchronously over WebSocket connection. 
+
+#### Send and receive messages using 'websocket:Client'
+
```

Comment:
Let's have a sub section for each remote service. Each subsection we can have signature of each remote method and the explain it.